File: sdk/erlang/src/atradio.erl
%% Official Erlang SDK for atradio.fm.
%%
%% A friendly wrapper over the Rustler NIF (`atradio_nif`) bindings to the shared
%% Rust core (`atradio-sdk`). Decodes the {"ok"|"error"} JSON envelopes with the
%% OTP `json` module (OTP 27+). The auth / record / reconcile logic is identical
%% to the Rust, Go, TypeScript, Python, Ruby, and Clojure SDKs.
%%
%% Stations are maps with binary camelCase keys
%% (#{<<"stationId">> => …, <<"name">> => …, <<"streamUrl">> => …, <<"source">> => …}).
-module(atradio).

-export([recent_stations/1, recent_stations/2,
         popular_stations/1, popular_stations/2,
         global_recently_played/1, global_recently_played/2,
         favorites/2, favorites/3,
         favorite_rkey/1,
         login/3, login/4,
         favorite/2, unfavorite/2, comment/3,
         set_play_status/2, delete_play_status/1, refresh_session/1]).

%% ---- reads (unauthenticated) ----

recent_stations(Limit) -> recent_stations(Limit, <<>>).
recent_stations(Limit, Base) ->
    unwrap(atradio_nif:recent_stations(bin(Base), Limit)).

popular_stations(Limit) -> popular_stations(Limit, <<>>).
popular_stations(Limit, Base) ->
    unwrap(atradio_nif:popular_stations(bin(Base), Limit)).

global_recently_played(Limit) -> global_recently_played(Limit, <<>>).
global_recently_played(Limit, Base) ->
    unwrap(atradio_nif:global_recently_played(bin(Base), Limit)).

favorites(Actor, Limit) -> favorites(Actor, Limit, <<>>).
favorites(Actor, Limit, Base) ->
    unwrap(atradio_nif:favorites(bin(Base), bin(Actor), Limit)).

%% The deterministic favorite record key — identical across every atradio SDK.
favorite_rkey(StationId) ->
    atradio_nif:favorite_rkey(bin(StationId)).

%% ---- authenticated agent ----
%% Returns an opaque resource handle; it's released automatically by GC.

login(Session, Id, Pw) -> login(Session, Id, Pw, <<>>).
login(Session, Id, Pw, AppView) ->
    atradio_nif:agent_login(bin(Session), bin(Id), bin(Pw), bin(AppView)).

favorite(Agent, Station) ->
    unwrap(atradio_nif:agent_favorite(Agent, json_bin(Station))).

unfavorite(Agent, Station) ->
    unwrap(atradio_nif:agent_unfavorite(Agent, json_bin(Station))).

comment(Agent, Station, Text) ->
    unwrap(atradio_nif:agent_comment(Agent, json_bin(Station), bin(Text))).

set_play_status(Agent, Station) ->
    unwrap(atradio_nif:agent_set_play_status(Agent, json_bin(Station))).

delete_play_status(Agent) ->
    unwrap(atradio_nif:agent_delete_play_status(Agent)).

refresh_session(Agent) ->
    unwrap(atradio_nif:agent_refresh_session(Agent)).

%% ---- internals ----

unwrap(Json) ->
    case json:decode(Json) of
        #{<<"error">> := E} -> error({atradio, E});
        #{<<"ok">> := Ok} -> Ok;
        Other -> Other
    end.

json_bin(Term) -> iolist_to_binary(json:encode(Term)).

bin(B) when is_binary(B) -> B;
bin(L) when is_list(L) -> list_to_binary(L);
bin(A) when is_atom(A) -> atom_to_binary(A, utf8).

File: sdk/elixir/lib/atradio.ex
defmodule Atradio do
  @moduledoc """
  Official Elixir SDK for atradio.fm.

  A thin wrapper over the `:atradio_erl` NIF package (the shared Rust core):
  AppView reads, record writes, and the deterministic favorite key — identical
  to the Rust, Go, TypeScript, Python, Ruby, and Clojure SDKs. `:atradio_erl`
  downloads the matching native library from the GitHub release on first use.

  Reads return lists/maps with binary keys (the wire shape). Stations passed to
  the write verbs are maps with binary keys — `"stationId"`, `"name"`,
  `"streamUrl"`, `"source"`.

      iex> Atradio.recent_stations(5) |> Enum.map(& &1["station"]["name"])
      iex> Atradio.favorite_rkey("rb:...")   # 16-char hex, matches every SDK
  """

  # ---- reads (unauthenticated) --------------------------------------------

  @doc "Newest stations platform-wide."
  def recent_stations(limit \\ 50, base \\ ""),
    do: :atradio.recent_stations(limit, to_bin(base))

  @doc "Most-favorited stations platform-wide."
  def popular_stations(limit \\ 50, base \\ ""),
    do: :atradio.popular_stations(limit, to_bin(base))

  @doc "Platform-wide who's-listening feed."
  def global_recently_played(limit \\ 50, base \\ ""),
    do: :atradio.global_recently_played(limit, to_bin(base))

  @doc "An actor's favorited stations."
  def favorites(actor, limit \\ 50, base \\ ""),
    do: :atradio.favorites(to_bin(actor), limit, to_bin(base))

  @doc "The deterministic favorite record key — identical across every atradio SDK."
  def favorite_rkey(station_id), do: :atradio.favorite_rkey(to_bin(station_id))

  # ---- authenticated agent -------------------------------------------------

  @doc """
  Log in with an app password, persisting the session at `session_path`.
  Returns an opaque agent handle (a resource freed by GC).
  """
  def login(session_path, identifier, password, appview \\ ""),
    do:
      :atradio.login(
        to_bin(session_path),
        to_bin(identifier),
        to_bin(password),
        to_bin(appview)
      )

  @doc "Favorite a station (idempotent; deterministic record key)."
  def favorite(agent, station), do: :atradio.favorite(agent, station)

  @doc "Unfavorite a station (removes every record for its stationId)."
  def unfavorite(agent, station), do: :atradio.unfavorite(agent, station)

  @doc "Post a comment on a station."
  def comment(agent, station, text), do: :atradio.comment(agent, station, to_bin(text))

  @doc "Update the actor's play-status singleton."
  def set_play_status(agent, station), do: :atradio.set_play_status(agent, station)

  @doc "Delete the actor's play-status singleton."
  def delete_play_status(agent), do: :atradio.delete_play_status(agent)

  @doc "Proactively refresh the session (keep-alive)."
  def refresh_session(agent), do: :atradio.refresh_session(agent)

  defp to_bin(s) when is_binary(s), do: s
  defp to_bin(s), do: to_string(s)
end
